src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -394,10 +418,16 @@ bool FileSegment::reserve(size_t size_to_reserve)
         expected_downloaded_size = getDownloadedSizeUnlocked(segment_lock);
 
         if (expected_downloaded_size + size_to_reserve > range().size())
-            throw Exception(
-                ErrorCodes::REMOTE_FS_OBJECT_CACHE_ERROR,
-                "Attempt to reserve space too much space ({}) for file segment with range: {} (downloaded size: {})",
-                size_to_reserve, range().toString(), downloaded_size);
+        {
+            if (strict)
+            {
+                throw Exception(
+                    ErrorCodes::REMOTE_FS_OBJECT_CACHE_ERROR,
+                    "Attempt to reserve space too much space ({}) for file segment with range: {} (downloaded size: {})",
+                    size_to_reserve, range().toString(), downloaded_size);
+            }
+            size_to_reserve = range().size() - expected_downloaded_size;
+        }
```

Comment:
But we cannot allow to reserve more than file segment size even if a certain flag is passes, it breaks the invariants, will make it difficult to write assertions. If you need a bigger size, then add required size to constructor of file segment.

Reply:
It doesn't allocate more than file segment size. In non strict mode we allocate as much space as possible, so it would reserve the rest space in current segment and returns amount of bytes we managed to reserve. I added comment to clarify it